test/e2e/snapshot_test.go

```diff
@@ -0,0 +1,404 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e_test
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"net/http/httptest"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+var fileIDs = map[string]map[string]int{}
+var memoryMap = map[string]map[string]any{}
+var lastData = map[string]map[string][]string{}
+
+func snapshotBaseDir(t *testing.T) string {
+	t.Helper()
+
+	dir, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if strings.HasSuffix(dir, "test/e2e") {
+		dir = path.Join(dir, ".snapshots")
+	} else {
+		dir = path.Join(dir, "test/e2e/.snapshots")
+	}
+
+	return dir
+}
+
+func snapshotDir(t *testing.T) string {
+	t.Helper()
+
+	testName, _ := splitTestName(t)
+
+	dir := snapshotBaseDir(t)
+
+	dir = path.Join(dir, testName)
+
+	return dir
+}
+
+func enforceDir(t *testing.T, filename string) {
+	t.Helper()
+
+	dir := path.Dir(filename)
+
+	if _, err := os.Stat(dir); err != nil {
+		if !os.IsNotExist(err) {
+			t.Fatal(err)
+		}
+
+		if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+func snapshotBaseName(t *testing.T, r *http.Request) string {
+	t.Helper()
+
+	dir := snapshotDir(t)
+
+	return fmt.Sprintf("%s/%s_%s", dir, r.Method, strings.ReplaceAll(strings.ReplaceAll(r.URL.Path, "/", "_"), ":", "_"))
+}
+
+func snapshotName(t *testing.T, r *http.Request) string {
+	t.Helper()
+
+	baseName := snapshotBaseName(t, r)
+
+	testName, _ := splitTestName(t)
+
+	if fileIDs[testName] == nil {
+		fileIDs[testName] = map[string]int{}
+	}
+	fileIDs[testName][baseName]++
+
+	id := fileIDs[testName][baseName]
+
+	fileName := fmt.Sprintf("%s_%d.json", baseName, id)
+
+	return fileName
+}
+
+func snapshotNameStepBack(t *testing.T, r *http.Request) {
+	t.Helper()
+
+	baseName := snapshotBaseName(t, r)
+
+	testName, _ := splitTestName(t)
+
+	if fileIDs[testName] == nil {
+		fileIDs[testName] = map[string]int{}
+	}
+	fileIDs[testName][baseName] -= 2
+	if fileIDs[testName][baseName] < 0 {
+		t.Fatal("no previous snapshot")
+	}
+}
+
+func updateSnapshots() bool {
+	return isTrue(os.Getenv("UPDATE_SNAPSHOTS"))
+}
+
+func skipSnapshots() bool {
+	return os.Getenv("UPDATE_SNAPSHOTS") == "skip"
+}
+
+func loadMemory(t *testing.T) {
+	t.Helper()
+
+	dir := snapshotDir(t)
+	filename := path.Join(dir, "memory.json")
+
+	buf, err := os.ReadFile(filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			memoryMap[t.Name()] = map[string]any{}
+			return
+		}
+		t.Fatal(err)
+	}
+
+	var data map[string]any
+	if err := json.Unmarshal(buf, &data); err != nil {
```

Comment:
[q] What's the reason for using a `map[string]map[string]any{}` over a `type Memory struct` that contains a strongly typed version of the memory?